engines/ags/shared/gfx/allegro_bitmap.cpp

```diff
@@ -123,13 +123,42 @@ void Bitmap::Destroy() {
 	_isDataOwner = false;
 }
 
+BITMAP *Bitmap::ConvertToAllegroFormat(BITMAP *al_bmp) {
+	Graphics::ManagedSurface *origSurface = new Graphics::ManagedSurface(al_bmp->getSurface());
+	Graphics::ManagedSurface *newSurface = nullptr;
+
+	if (origSurface->format.bytesPerPixel == 2 && origSurface->format != Graphics::PixelFormat(2, 5, 6, 5, 0, 11, 5, 0, 0)) { // convert to RGB565
+		newSurface = new Graphics::ManagedSurface(origSurface->w, origSurface->h, Graphics::PixelFormat(2, 5, 6, 5, 0, 11, 5, 0, 0));
+	} else if ((origSurface->format.bytesPerPixel == 3) ||
+			   (origSurface->format.bytesPerPixel == 4 && origSurface->format != Graphics::PixelFormat(4, 8, 8, 8, 8, 16, 8, 0, 24))) { // convert to RGBA8888
+		newSurface = new Graphics::ManagedSurface(origSurface->w, origSurface->h, Graphics::PixelFormat(4, 8, 8, 8, 8, 16, 8, 0, 24));
+	} else { // no conversion needed
+		origSurface->free();
+		delete origSurface;
+		return al_bmp;
+	}
+
+	if (newSurface) {
+		newSurface->blitFrom(*origSurface);
+		BITMAP *conv_al_bmp = new BITMAP(newSurface);
+		origSurface->free();
+		delete origSurface;
+		return conv_al_bmp;
+	}
+	origSurface->free();  // failed conversion
+	delete origSurface;
+	return nullptr;
+}
+
 bool Bitmap::LoadFromFile(const char *filename) {
 	Destroy();
 
 	BITMAP *al_bmp = load_bitmap(filename, nullptr);
 	if (al_bmp) {
-		_alBitmap = al_bmp;
-		_isDataOwner = true;
+		_alBitmap = ConvertToAllegroFormat(al_bmp);
+		destroy_bitmap(al_bmp);
```

Comment:
You can't do this: if no conversion is needed, `al_bmp` is returned and will be freed: `_alBitmap` will point to freed memory.
You can free `al_bmp` directly in the helper function when converting or check here `if (_alBitmap != al_bmp)` and delete `al_bmp` in this case.
I find this less clean but it's personal opinion.